File: SegundaParte/Semana8/C++/ExSingleton.cpp
#include <iostream>
#include <vector>
#include <string>
#include <typeinfo>

class Log {
private:
    int cont;
    std::vector<std::string> log;

    Log() : cont(0) {}
public:
    static Log& getInstance() {
        static Log instance;
        return instance;
    }

    void insereAtividade(const std::string& className) {
        cont++;
        log.push_back(std::to_string(cont) + " & " + className);
    }

    void printLog() {
        std::cout << "Log de criação de objetos:" << std::endl;
        for (const auto& entry : log) {
            std::cout << entry << std::endl;
        }
    }
};

class XXX {
public:
    XXX() {
	Log &inst = Log::getInstance();
        inst.insereAtividade(typeid(*this).name());
    }
};

class YYY {
public:
    YYY() {
	Log &inst = Log::getInstance();
        inst.insereAtividade(typeid(*this).name());
    }
};

class ZZZ {
public:
    ZZZ() {
	Log &inst = Log::getInstance();
        inst.insereAtividade(typeid(*this).name());
    }
};

int main() {
    XXX objetoXXX1;
    XXX objetoXXX2;
    YYY objetoYYY;
    XXX objetoXXX3;
    ZZZ objetoZZZ;

    Log::getInstance().printLog();

    return 0;
}

File: SegundaParte/Semana8/Java/ExSingleton.java
import java.util.ArrayList;
import java.util.List;

class Log {
    private int cont;
    private List<String> log;
    static Log instance;

    static { instance = new Log(); }

    private Log() {
        cont = 0;
        log = new ArrayList<>();
    }

    public static Log getInstance() {
        return instance;
    }

    public void insereAtividade(String className) {
        cont++;
        log.add(cont + " & " + className);
    }

    public void printLog() {
        System.out.println("Log de criação de objetos:");
        for (String entry : log) {
            System.out.println(entry);
        }
    }
}

class XXX {
    public XXX() {
	Log inst = Log.getInstance();
        inst.insereAtividade(getClass().getName());
    }
}

class YYY {
    public YYY() {
	Log inst = Log.getInstance();
        inst.insereAtividade(getClass().getName());
    }
}

class ZZZ {
    public ZZZ() {
	Log inst = Log.getInstance();
        inst.insereAtividade(getClass().getName());
    }
}

public class ExSingleton {
    public static void main(String[] args) {
        XXX objetoXXX1 = new XXX();
        XXX objetoXXX2 = new XXX();
        YYY objetoYYY = new YYY();
        XXX objetoXXX3 = new XXX();
        ZZZ objetoZZZ = new ZZZ();

        Log.getInstance().printLog();
    }
}

